Lib/collections/__init__.py

```diff
@@ -390,46 +340,109 @@ def namedtuple(typename, field_names, *, verbose=False, rename=False, module=Non
                 or _iskeyword(name)
                 or name.startswith('_')
                 or name in seen):
-                field_names[index] = '_%d' % index
+                field_names[index] = f'_{index}'
             seen.add(name)
     for name in [typename] + field_names:
         if type(name) is not str:
             raise TypeError('Type names and field names must be strings')
         if not name.isidentifier():
             raise ValueError('Type names and field names must be valid '
-                             'identifiers: %r' % name)
+                             f'identifiers: {name!r}')
         if _iskeyword(name):
             raise ValueError('Type names and field names cannot be a '
-                             'keyword: %r' % name)
+                             f'keyword: {name!r}')
     seen = set()
     for name in field_names:
         if name.startswith('_') and not rename:
             raise ValueError('Field names cannot start with an underscore: '
-                             '%r' % name)
+                             f'{name!r}')
         if name in seen:
-            raise ValueError('Encountered duplicate field name: %r' % name)
+            raise ValueError('Encountered duplicate field name: {name!r}')
         seen.add(name)
 
-    # Fill-in the class template
-    class_definition = _class_template.format(
-        typename = typename,
-        field_names = tuple(field_names),
-        num_fields = len(field_names),
-        arg_list = repr(tuple(field_names)).replace("'", "")[1:-1],
-        repr_fmt = ', '.join(_repr_template.format(name=name)
-                             for name in field_names),
-        field_defs = '\n'.join(_field_template.format(index=index, name=name)
-                               for index, name in enumerate(field_names))
+    # Variables used in the methods and docstrings
+    field_names = tuple(field_names)
+    num_fields = len(field_names)
+    arg_list = repr(field_names).replace("'", "")[1:-1]
+    repr_fmt = '(' + ', '.join(f'{name}=%r' for name in field_names) + ')'
+    module_name = f'namedtuple_{typename}'
+    tuple_new = tuple.__new__
+    _len = len
+
+    # Create all the named tuple methods to be added to the class namespace
+
+    s = f'def __new__(_cls, {arg_list}): return _tuple_new(_cls, ({arg_list}))'
+    namespace = dict(_tuple_new=tuple_new, __name__=module_name)
```

Comment:
Could you explain why you chose dict() over {k:v}? I believe {k:v} is faster because it doesn't need to make a function call.
Could you explain why you chose dict() over {k:v}? I believe {k:v} is faster because it doesn't need to make a function call.

Reply:
Beauty counts.

Reply:
Fair enough. Now that I think about it, I wonder if optimizing it in peephole would be a good thing? The pattern is pretty consistent `LOAD_GLOBAL` (dict) (`LOAD_CONST` arg names and values) `CALL_FUNCTION_KW`.

Reply:
Can't do it because __builtins__.dict can be replaced by a user or shadowed in globals().  Just the facts of life in a very dynamic language.

Reply:
Of course. Thanks for the explanation! I should have guessed there was good reason for a "simple improvement" such as that not being implemented.